src/data/equiv/algebra.lean

```diff
@@ -355,4 +380,112 @@ instance symm.is_ring_hom {e : α ≃r β} : is_ring_hom e.to_equiv.symm := hom
+  {f g : ring_equiv R S} (h : f.to_fun = g.to_fun) : f = g :=
+begin
+  have h₁ := @equiv.eq_of_to_fun_eq _ _ f.to_equiv g.to_equiv h,
+  cases f, cases g, congr,
+  { exact h },
+  { exact congr_arg equiv.inv_fun h₁ }
+end
+
 end ring_equiv
+
+/-
+We define the type of automorphisms on groups, additive monoids,
+additive groups and rings using `mul_equiv`, `add_equiv` and `ring_equiv`.
+In each case this type also forms a group.
+-/
+
+namespace monoid
+
+def aut (γ : Type u) [has_mul γ] := mul_equiv γ γ
+
+/--
+The group operation on automorphisms of a monoid is defined by
+λ g h, mul_equiv.trans h g.
+This means that multiplication agrees with composition, (g*h)(x) = g (h x) .
+-/
+instance aut_group (γ : Type u) [has_mul γ] : group (aut γ) :=
+{ mul := λ g h, mul_equiv.trans h g,
+  one := mul_equiv.refl γ,
+  inv := mul_equiv.symm,
+  mul_assoc := λ _ _ _, by { ext, refl },
+  one_mul := λ _, by { ext, refl },
+  mul_one := λ _, by { ext, refl },
+  mul_left_inv := λ _, by { ext, apply equiv.left_inv } }
+
+end monoid
+
+namespace group
+
+-- A group homomorphism is a monoid homomorphism between groups.
+def aut (γ : Type u) [group γ] := mul_equiv γ γ
+
+instance aut_group (γ : Type u) [group γ] : group (aut γ) := monoid.aut_group γ
+
+end group
+
+namespace add_monoid
+
+def aut (α : Type u) [has_add α] := add_equiv α α
+
+/--
+The group operation on automorphisms of an additive monoid is defined by
+λ g h, add_equiv.trans h g.
+This means that multiplication agrees with composition, (g*h)(x) = g (h x) .
+-/
+instance aut_group (α : Type u) [has_add α] : group (aut α) :=
+{ mul := λ g h, add_equiv.trans h g,
+  one := add_equiv.refl α,
+  inv := add_equiv.symm,
+  mul_assoc := λ _ _ _, by { ext, refl },
+  one_mul := λ _, by { ext, refl },
+  mul_one := λ _, by { ext, refl },
+  mul_left_inv := λ _, by { ext, apply equiv.left_inv } }
+
+end add_monoid
+
+namespace add_group
+
+-- An additive group homomorphism is an additive monoid homomorphism between groups.
+def aut (γ : Type u) [add_group γ] := add_equiv γ γ
+
+instance aut_group (γ : Type u) [add_group γ] : group (aut γ) := add_monoid.aut_group γ
+
+end add_group
+
+namespace ring
+
+def aut (R : Type u) [ring R] := ring_equiv R R
+
+/--
+The group operation on automorphisms of a ring is defined by
+λ g h, ring_equiv.trans h g.
+This means that multiplication agrees with composition, (g*h)(x) = g (h x) .
+-/
+instance aut_group (R : Type u) [ring R] : group (aut R) :=
+{ mul := λ g h, ring_equiv.trans h g,
+  one := ring_equiv.refl R,
+  inv := ring_equiv.symm,
+  mul_assoc := λ _ _ _, by { ext, refl },
+  one_mul := λ _, by { ext, refl },
+  mul_one := λ _, by { ext, refl },
+  mul_left_inv := λ _, by { ext, apply equiv.left_inv } }
+
+/--
+Map from the automorphisms of a ring to the automorphisms of the additive
+group defined by addition in R.
+-/
+def ring_aut_to_add_monoid_aut (R : Type u) [ring R] (f : aut R) : add_monoid.aut R :=
```

Comment:
Why do you go to `add_monoid.aut`? Why not to `add_group.aut`? Also, there are maps to `group.aut` and `monoid.aut`...

Reply:
I changed this to `add_group.aut`, but I'm not sure of the best structure to handle all of these forgetful maps. Perhaps ... we should chuck this all out in favour of using the categorical `Aut`... :-)

Reply:
I thought that the main reason to have `mul_equiv` etc (apart from different universes) is to be able to write `α ≃* β` and do not care if `α` and `β` are (commutative) monoids or groups. Why not use the same approach with `aut`s?

Reply:
Defined bundled homs from `ring_aut` to `mul_aut`, `add_aut`, and `equiv.perm`. Unfortunately, this means that we can't say `f.to_mul_aut`.